### DISCUSSION TITLE
Two-way communication Home Assistant Arduino mega + Tasmota

### DISCUSSION THREAD

**Discussion by GeorgeWatch** [arendst/Tasmota#15285] [Q&A] [answered]
Hi,
I solve how to communicate with Home Assistant using Tasmota and Mega 2560. I send the following to HA using Mega2560 temp sensor / Tasmota:
Serial2.print( "{\"B00\":"+b[1]+",\"B01\":"+b[2]+",\"B02\":"+b[3]+",\"B03\":"+b[4]+",\"B04\":"+b[5]+",\"B05\":"+b[6]+",\"B06\":"+b[7]+",\"B07\":"+b[8]+ "}");

Tasmota console displays: 
19:05:35 MQT: tele/TasmotaPLC/RESULT = {"SerialReceived":{"B00":382.2,"B01":382.2,"B02":382.2,"B03":382.2,"B04":382.2,"B05":382.2,"B06":1.9,"B07":0.2}}

HA will handle it.

When I send HA data using Node-RED: 
cmnd/TasmotaPLC/**SerialSend**
sensor.boiler_localtuya_w

Node-RED Debug:
sensor.boiler_localtuya_w : msg : Object
{ payload: "235.6", data: "sensor.boiler_localtuya_w“, topic: "sensor.boiler_localtuya_w“, _msgid: "02867b5d3724fdfd" } 

Tasmota console displays: 
19:05:39 MQT: stat/TasmotaPLC/RESULT = {"SerialSend":"Done"}
MQTT explorer:
cmnd/TasmotaPLC/SerialSend =6687.35

How do I send data from HA to Mega 2560? 

Thank you for your help.

**Comment by barbudor** [comment, marked as answer]
As you have found how to send data to Arduino from Nodered, you have to do the same from HA
Better get help from a HA forum

Note that if it is difficult to send a "SerialSend" formatted as you want from HA, you may use Rules to format on Tasmota
For exemple 
```
Rule1
  ON event#config DO SerialSend { payload: "%value%", data: "sensor.boiler_localtuya_w“, topic: "sensor.boiler_localtuya_w“, _msgid: "02867b5d3724fdfd" } ENDON
```
Then a simple publish on `cmnd/yourtopic/event` with payload `config=235.6` would do as your above example

**Reply by GeorgeWatch** [reply]
Thank you for answer. The confusing thing is that the console does not display SerialSend content but "Done" 
I'll try to google an example. I don't have much experience with HA 
GW

**Reply by barbudor** [reply]
The content sent over serial is the payload you created in NodeRed for the SerialSend command. There's no reasons why the acknowledge of the command should echo the data sent.
You should be able to see an echo of the data for debugging by rising WebLog level to 3

**Reply by GeorgeWatch** [reply]
Thanks, I guess it works using a feature that wraps it in one message.

HA Node-RED
FUNCTION:
bookmark:  On Message

msg.payload = {
   topic: msg.topic,
   payload: msg.payload,
   // etc ...
}
return msg; 

Tasmota console Weblog level 3:
12:55:10 CMD: Group 0, Index 1, Command "SERIALSEND", Data "{"topic":"sensor.elektrokotel_localtuya_total_energy","payload":"6693.67"}"
12:54:46 MQT: stat/TasmotaPLC/RESULT = {"SerialSend":"Done"}
12:54:48 SRC: MQTT
12:54:48 CMD: Group 0, Index 1, Command "SERIALSEND", Data "{"topic":"sensor.elektrokotel_localtuya_prikon","payload":"3050.3"}"
12:54:48 MQT: stat/TasmotaPLC/RESULT = {"SerialSend":"Done"}

I haven't come up with anything better yet  :-D

![Clipboard04](https://user-images.githubusercontent.com/46403712/161536990-b2420fc4-ef2e-477f-8d5e-8ed800ad03e6.jpg)

**Reply by barbudor** [reply]
Node-RED is a nice tool to bridge different systems 👍

**Reply by GeorgeWatch** [reply]
Yes it's good. Topic seems long to me. I rewrote it for short topic :-D 
--------------------                                                                                      ------------------------------------
var text = msg.topic 
msg.payload = {
  topic: text.replace("sensor.elektrokotel_localtuya","elk"),
  payload: msg.payload
}
return msg;
--------------------------------------------------------------------------------------------------------------------
Tasmota:
14:06:10 CMD: Group 0, Index 1, Command "SERIALSEND", Data "{"topic":"elk_prikon","payload":"3031.4"}"

**Comment by ThomasSBW** [comment]
Dear all,

I am fighting with the Serial bridge several days to find a solution to communicate between Arduino <--> Tasmota Bridge <--> Mosquitto MQTT.
After I assemble the Bin file I was able to flash it on ESP01s.
I configure the SerTX/RX to GPIO1, 3 enable the MQTT and setup the MQTT settings.
Serial Port Pins I connect to a USB Serial Adapter for the first testing.
Before my Systems was working more standallone but I try to start so bring them more and more to a home network and about this also my expiriance with MQTT is also not very strong.
Anyway after setup the ESP01s with Tasmota Bridge I start to send my JSON Payload by serial Terminal to the ESP and what I should say it worked immedially but then my nightmare starts with the communication MQTT to ESP.

Communication Serial to ESP:
`21:26:07.154 MQT: serial/serialbridge/RESULT = {"SSerialReceived":{"Sensoren":{"Anzahl":"3","Sensor01":{"ID":"77777777","Temperature": "21.6"},"Sensor02":{"ID":"87777777","Temperature": "22.6"},"Sensor03":{"ID":"88777777","Temperature": "28.6"},"TempUnit": "C"}}}`

Mostly this are the data I would like to store in a Database and this worked already with Node Red (I also start 2 weeks ago) so not so bad.
The only Point is the communication backward and I have really no idea how I can manage this after a week.
I use MQTT Explorer to publish a message to the ESP but I receive nothing on the ESP Console I read many Tasmoty sides but maybe I cannot understand or not correctly.
Maybe also the Programmer gen is not very high so maybe I understand  it wrong.
For me the follow points are unclear for me maybe more.

1. I think a rule is nessesarry.
2. Correct publish Topic I think with cnmd but I am not sure where and if I need it.
3. Setup the Config on the ESP or not?

The Command will always the same format only with different values.
Example:
`{
 "Command":{
  "Device":{"ID":"123456789","Job": "Del"},
 }
}`

Hope somebody can halp me with this problem.
Thanks in advance

Thomas

PS: sorry for my english.

**Reply by barbudor** [reply]
First, it looks like you have broken your FullTopic
You MUST have `%prefix%`  and ` %topic%`  somewhere is your FullTopic which is obvioulsy not the case as the topic you are publishing to is obvioulsy wrong. You can re-order, add other topic level, but you MUST have both
Without respecting the FullTopic structure you will never be able to send a command to your Tasmota

If you want the command to be sent to the Arduino, you should send to Tasmota a `SerialSend` command
WIth a standard Fulltopic as `%prefix%/%topic%` you have to publish to topic `cmnd/serialbridge/SerialSend` and the payload would be your text to be send to your Arduino
There are different variant of SerialSend (SerialSend1 which adds an \n after the message, SerialSend2 which doesn't, ....

By the way, did you looked into https://tasmota.github.io/docs/TasmotaClient/ which offer a standarized way of communication between an Arduino and Tasmota ? It also support software download to the Atmel CPU (if properly connected)

**Reply by ThomasSBW** [reply]
Dear barbudor,
first thanks for answering.
About the FullTopic a play a lot so I forget to send the Prefix back.
Here a new Output with the "original" Settings.
`22:18:39.425 MQT: serial/tele/serialbridge_51FE95/STATE = {"Time":"2023-01-10T22:18:39","Uptime":"0T00:00:10","UptimeSec":10,"Heap":27,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Wifi":{"AP":1,"SSId":"XxXxXxXx","BSSId":"XX:9X:XX:XX:XX:XX","Channel":11,"Mode":"11n","RSSI":72,"Signal":-64,"LinkCount":1,"Downtime":"0T00:00:04"}}
22:18:56.119 MQT: serial/tele/serialbridge_51FE95/RESULT = {"SSerialReceived":{"Sensoren":{"Anzahl":"3","Sensor01":{"ID":"77777777","Temperature": "21.6"},"Sensor02":{"ID":"87777777","Temperature": "22.6"},"Sensor03":{"ID":"88777777","Temperature": "28.6"},"TempUnit": "C"}}}
22:19:08.191 MQT: serial/tele/serialbridge_51FE95/RESULT = {"SSerialReceived":{"Sensoren":{"Anzahl":"3","Sensor01":{"ID":"77777777","Temperature": "21.6"},"Sensor02":{"ID":"87777777","Temperature": "22.6"},"Sensor03":{"ID":"88777777","Temperature": "28.6"},"TempUnit": "C"}}}`

When is setup publish topic `cmnd/tele/serialbridge_51FE95/SerialSend` and the payload `{"Command":{"ID":"123456789","Job":"Del"}}`.
I see the measage in the MQTT Exporter but not in the Tasmota Console. --> see the attached Pictures.
Also in the Serial Terminal I can see some income statements by ESP booting up so the communication looks OK.

The Link page I didn't saw until now but I will check.

Thanks in advance

Thomas
![Settings](https://user-images.githubusercontent.com/122159548/211671176-e1ec39f9-1289-40d0-8a27-de1cb432f59e.png)
![Output](https://user-images.githubusercontent.com/122159548/211671179-11a67aff-aba8-4ef5-91cc-bbd774001cf4.png)

**Reply by barbudor** [reply]
To properly issue a command, the keyword `cmnd` should take the place of ` %prefix%` 
So FullTopic `serial/%prefix/%topic%`  with `%topic%` being `serialbridge_51FE95` mean that you should send to `serial/cmnd/serialbridge_51FE95/SerialSend`

**Reply by ThomasSBW** [reply]
Dear barbudor,
I try your settings and it worked.
May thanks that you bring me one step forward.

Thanks

Thomas

**Comment by Gifford47** [comment]
@GeorgeWatch I also want to use tasmota with an arduino 2560 mcu. Does OTA already works with a 2560 MCU?
